Nomad crashes when submitting job "runtime error: index out of range [0] with length 0"

Our company is evaluating using nomad, however nomad frequently crashes when a user submits a job. 
The main error is:

```
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]:     2020-09-11T17:28:53.418Z [ERROR] http: http: panic serving 127.0.0.1:35224: runtime error: index out of range [0] with length 0
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]: goroutine 157 [running]:
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]: net/http.(*conn).serve.func1(0xc000e92640)
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]:         net/http/server.go:1800 +0x139
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]: panic(0x2f79900, 0xc001172c00)
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]:         runtime/panic.go:975 +0x3e3
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]: 
```

## Version
0.12.4


<details><summary>Nomad Job</summary>
<p>

```
job "pulse-predictions-service-mm-nomad-and-roam" {
  region = "sand"
  datacenters = ["sb"]
  type = "service"
  spread {
    attribute = "${node.datacenter}"
    weight = 100
  }
  constraint {
    attribute = "${meta.mx_app}"
    value = "common"
  }
  group "group" {
    count = 1    
    service { 
      name = "foobar"
      tags = [ "foobar" ]
      port = "http"
      check {
        name = "foobar-health"
        type = "http"
        path = "/"
        interval = "10s"
        timeout = "2s"
        expose   = true
      }
      connect {
        sidecar_service {}
      }
    }
    
    task "app" {
      driver = "docker"
      config {
        image = "gitlab.example.com/redacted"
        volumes = []
      }
      env {
        APP_NAME = "foobar"
        APP_HOST = "${attr.unique.network.ip-address}"
      }
      resources {
        cpu    = 250
        memory = 512
        network {
            mbits = 10
        }
      }
    } # task
  } # group
  meta {
    CI_PROJECT_URL = "https://gitlab.example.com/redacted"
    CI_COMMIT_REF_NAME = "redacted"
    EPHEMERAL = "true"
    CI_PIPELINE_IID = "14"
    CI_COMMIT_SHA   = "redacted"
  }
} # job
```
</p>
</details>






<details><summary>Server Logs</summary>
<p>

```
Sep 11 17:28:33 sb-sand-nomad2 nomad[6188]:     2020-09-11T17:28:33.594Z [INFO]  nomad: serf: EventMemberJoin: sb-sand-nomad1.sand 10.47.69.194
Sep 11 17:28:33 sb-sand-nomad2 nomad[6188]:     2020-09-11T17:28:33.594Z [INFO]  nomad: serf: EventMemberJoin: sb-sand-nomad3.sand 10.47.69.196
Sep 11 17:28:33 sb-sand-nomad2 nomad[6188]:     2020-09-11T17:28:33.594Z [INFO]  nomad: serf: Re-joined to previously known node: sb-sand-nomad1.sand: 10.47.69.194:4648
Sep 11 17:28:33 sb-sand-nomad2 nomad[6188]:     2020-09-11T17:28:33.594Z [INFO]  nomad: adding server: server="sb-sand-nomad1.sand (Addr: 10.47.69.194:4647) (DC: sb)"
Sep 11 17:28:33 sb-sand-nomad2 nomad[6188]:     2020-09-11T17:28:33.594Z [INFO]  nomad: adding server: server="sb-sand-nomad3.sand (Addr: 10.47.69.196:4647) (DC: sb)"
Sep 11 17:28:33 sb-sand-nomad2 nomad[6188]:     2020-09-11T17:28:33.694Z [INFO]  nomad.vault: successfully renewed token: next_renewal=35h59m59.999993572s
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]:     2020-09-11T17:28:53.418Z [ERROR] http: http: panic serving 127.0.0.1:35224: runtime error: index out of range [0] with length 0
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]: goroutine 157 [running]:
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]: net/http.(*conn).serve.func1(0xc000e92640)
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]:         net/http/server.go:1800 +0x139
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]: panic(0x2f79900, 0xc001172c00)
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]:         runtime/panic.go:975 +0x3e3
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]: github.com/hashicorp/nomad/nomad.exposePathForCheck(0xc001127000, 0xc0009b24d0, 0xc001165b00, 0xc00112d440, 0x0, 0x0)
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]:         github.com/hashicorp/nomad/nomad/job_endpoint_hook_expose_check.go:213 +0x59e
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]: github.com/hashicorp/nomad/nomad.jobExposeCheckHook.Mutate(0xc00112d440, 0xc00155b060, 0x411858, 0x60, 0x2965820, 0x0, 0x0)
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]:         github.com/hashicorp/nomad/nomad/job_endpoint_hook_expose_check.go:27 +0x8e
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]: github.com/hashicorp/nomad/nomad.(*Job).admissionMutators(0xc0004bab40, 0xc00112d440, 0x7, 0xc0009def01, 0xc00155b368, 0xc001325140, 0xc00062eb90, 0x2d62640)
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]:         github.com/hashicorp/nomad/nomad/job_endpoint_hooks.go:61 +0xd3
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]: github.com/hashicorp/nomad/nomad.(*Job).Validate(0xc0004bab40, 0xc000af2bc0, 0xc000af2c00, 0x0, 0x0)
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]:         github.com/hashicorp/nomad/nomad/job_endpoint.go:537 +0x382
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]: reflect.Value.call(0xc000234240, 0xc000010ab0, 0x13, 0x31f31a9, 0x4, 0xc00155b698, 0x3, 0x3, 0xc000af2c00, 0xc00155b6f0, ...)
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]:         reflect/value.go:460 +0x8ab
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]: reflect.Value.Call(0xc000234240, 0xc000010ab0, 0x13, 0xc00155b698, 0x3, 0x3, 0x8, 0x16, 0x16)
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]:         reflect/value.go:321 +0xb4
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]: net/rpc.(*service).call(0xc00040fe40, 0xc0004ba960, 0xc0003c5300, 0x0, 0xc00017f400, 0xc000c38040, 0x2f85b60, 0xc000af2bc0, 0x16, 0x2ccd1c0, ...)
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]:         net/rpc/server.go:377 +0x17f
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]: net/rpc.(*Server).ServeRequest(0xc0004ba960, 0x38c6e00, 0xc000af2b80, 0xc0012242c8, 0xc00006ca80)
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]:         net/rpc/server.go:498 +0x23e
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]: github.com/hashicorp/nomad/nomad.(*Server).RPC(0xc000503b00, 0x3203c7c, 0xc, 0x2f85b60, 0xc000af2b00, 0x2ccd1c0, 0xc000af2b40, 0xc001126f01, 0xc000af2b40)
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]:         github.com/hashicorp/nomad/nomad/server.go:1558 +0xb9
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]: github.com/hashicorp/nomad/command/agent.(*Agent).RPC(0xc0002fc7e0, 0x3203c7c, 0xc, 0x2f85b60, 0xc000af2b00, 0x2ccd1c0, 0xc000af2b40, 0xc000bae958, 0x65cb1d)
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]:         github.com/hashicorp/nomad/command/agent/agent.go:977 +0x7d
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]: github.com/hashicorp/nomad/command/agent.(*HTTPServer).ValidateJobRequest(0xc0007904b0, 0x38ab540, 0xc001326d20, 0xc001126f00, 0x4d8096, 0x5f5bb3d5, 0x18d83e15, 0xbf041b82a426f)
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]:         github.com/hashicorp/nomad/command/agent/job_endpoint.go:190 +0x1de
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]: github.com/hashicorp/nomad/command/agent.(*HTTPServer).wrap.func1(0x38ab540, 0xc001326d20, 0xc001126f00)
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]:         github.com/hashicorp/nomad/command/agent/http.go:448 +0x176
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]: net/http.HandlerFunc.ServeHTTP(0xc0007f5720, 0x38ab540, 0xc001326d20, 0xc001126f00)
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]:         net/http/server.go:2041 +0x44
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]: net/http.(*ServeMux).ServeHTTP(0xc0007cfdc0, 0x38ab540, 0xc001326d20, 0xc001126f00)
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]:         net/http/server.go:2416 +0x1a5
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]: github.com/NYTimes/gziphandler.GzipHandlerWithOpts.func1.1(0x38b3bc0, 0xc0004cd180, 0xc001126f00)
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]:         github.com/NYTimes/gziphandler@v1.0.1/gzip.go:277 +0x1e6
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]: net/http.HandlerFunc.ServeHTTP(0xc000c880c0, 0x38b3bc0, 0xc0004cd180, 0xc001126f00)
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]:         net/http/server.go:2041 +0x44
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]: net/http.serverHandler.ServeHTTP(0xc0007a7180, 0x38b3bc0, 0xc0004cd180, 0xc001126f00)
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]:         net/http/server.go:2836 +0xa3
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]: net/http.(*conn).serve(0xc000e92640, 0x38c6000, 0xc000af2100)
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]:         net/http/server.go:1924 +0x86c
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]: created by net/http.(*Server).Serve
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]:         net/http/server.go:2962 +0x35c
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]:     2020-09-11T17:28:53.908Z [ERROR] worker: failed to dequeue evaluation: error="rpc error: EOF"
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]:     2020-09-11T17:28:53.908Z [ERROR] worker: failed to dequeue evaluation: error="rpc error: EOF"
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]:     2020-09-11T17:28:53.930Z [ERROR] worker: failed to dequeue evaluation: error="rpc error: failed to get conn: dial tcp 10.47.69.194:4647: connect: connection refused"
Sep 11 17:28:53 sb-sand-nomad2 nomad[6188]:     2020-09-11T17:28:53.930Z [ERROR] worker: failed to dequeue evaluation: error="rpc error: failed to get conn: rpc error: lead thread didn't get connection"
Sep 11 17:28:54 sb-sand-nomad2 nomad[6188]:     2020-09-11T17:28:54.011Z [ERROR] worker: failed to dequeue evaluation: error="rpc error: failed to get conn: dial tcp 10.47.69.194:4647: connect: connection refused"
Sep 11 17:28:54 sb-sand-nomad2 nomad[6188]:     2020-09-11T17:28:54.011Z [ERROR] worker: failed to dequeue evaluation: error="rpc error: failed to get conn: rpc error: lead thread didn't get connection"
Sep 11 17:28:54 sb-sand-nomad2 nomad[6188]:     2020-09-11T17:28:54.011Z [ERROR] worker: failed to dequeue evaluation: error="rpc error: failed to get conn: rpc error: lead thread didn't get connection"
Sep 11 17:28:54 sb-sand-nomad2 nomad[6188]:     2020-09-11T17:28:54.333Z [ERROR] worker: failed to dequeue evaluation: error="rpc error: failed to get conn: dial tcp 10.47.69.194:4647: connect: connection refused"
Sep 11 17:28:54 sb-sand-nomad2 nomad[6188]:     2020-09-11T17:28:54.333Z [ERROR] worker: failed to dequeue evaluation: error="rpc error: failed to get conn: rpc error: lead thread didn't get connection"
Sep 11 17:28:54 sb-sand-nomad2 nomad[6188]:     2020-09-11T17:28:54.333Z [ERROR] worker: failed to dequeue evaluation: error="rpc error: failed to get conn: rpc error: lead thread didn't get connection"
Sep 11 17:28:55 sb-sand-nomad2 nomad[6188]:     2020-09-11T17:28:55.275Z [WARN]  nomad.raft: heartbeat timeout reached, starting election: last-leader=10.47.69.194:4647
Sep 11 17:28:55 sb-sand-nomad2 nomad[6188]:     2020-09-11T17:28:55.275Z [INFO]  nomad.raft: entering candidate state: node="Node at 10.47.69.195:4647 [Candidate]" term=34763
Sep 11 17:28:55 sb-sand-nomad2 nomad[6188]:     2020-09-11T17:28:55.294Z [ERROR] nomad.raft: failed to make requestVote RPC: target="{Voter 10.47.69.194:4647 10.47.69.194:4647}" error="dial tcp 10.47.69.194:4647: connect: connection refused"
```
```
</p>
</details>

--- a/nomad/job_endpoint_hook_expose_check.go
+++ b/nomad/job_endpoint_hook_expose_check.go
@@ -170,7 +170,7 @@ func serviceUsesConnectEnvoy(s *structs.Service) bool {
 	}
 
 	// A non-nil connect.sidecar_task stanza implies the sidecar task is being
-	// overridden (i.e. the default Envoy is not being uesd).
+	// overridden (i.e. the default Envoy is not being used).
 	if s.Connect.SidecarTask != nil {
 		return false
 	}
@@ -197,6 +197,12 @@ func checkIsExposable(check *structs.ServiceCheck) bool {
 func exposePathForCheck(tg *structs.TaskGroup, s *structs.Service, check *structs.ServiceCheck) (*structs.ConsulExposePath, error) {
 	if !checkIsExposable(check) {
 		return nil, nil
+	}
+
+	// Borrow some of the validation before we start manipulating the group
+	// network, which needs to exist once.
+	if err := tgValidateUseOfBridgeMode(tg); err != nil {
+		return nil, err
 	}
 
 	// If the check is exposable but doesn't have a port label set build

--- a/nomad/job_endpoint_hook_expose_check_test.go
+++ b/nomad/job_endpoint_hook_expose_check_test.go
@@ -376,6 +376,30 @@ func TestJobExposeCheckHook_exposePathForCheck(t *testing.T) {
 			ListenerPort:  tg.Networks[0].DynamicPorts[0].Label,
 		}, ePath)
 	})
+
+	t.Run("missing network with no service check port label", func(t *testing.T) {
+		// this test ensures we do not try to manipulate the group network
+		// to inject an expose port if the group network does not exist
+		c := &structs.ServiceCheck{
+			Name:      "check1",
+			Type:      "http",
+			Path:      "/health",
+			PortLabel: "",   // not set
+			Expose:    true, // will require a service check port label
+		}
+		s := &structs.Service{
+			Name:   "service1",
+			Checks: []*structs.ServiceCheck{c},
+		}
+		tg := &structs.TaskGroup{
+			Name:     "group1",
+			Services: []*structs.Service{s},
+			Networks: nil, // not set, should cause validation error
+		}
+		ePath, err := exposePathForCheck(tg, s, c)
+		require.EqualError(t, err, `group "group1" must specify one bridge network for exposing service check(s)`)
+		require.Nil(t, ePath)
+	})
 }
 
 func TestJobExposeCheckHook_containsExposePath(t *testing.T) {

